pkg/retainer/txretainer/tx_retainer_cache.go

```diff
@@ -0,0 +1,217 @@
+package txretainer
+
+import (
+	"sync"
+
+	"github.com/iotaledger/hive.go/ds/shrinkingmap"
+	"github.com/iotaledger/hive.go/ierrors"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/hive.go/runtime/options"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+// transactionRetainerCache is a cache for transaction metadata changes that are not yet committed to the database.
+type transactionRetainerCache struct {
+	accessMutex sync.RWMutex
+
+	txMetadataUpdateFunc txMetadataUpdateFunc
+
+	uncommittedTxMetadataChangesByID *shrinkingmap.ShrinkingMap[iotago.TransactionID, *TransactionMetadata]
+	uncommittedTxIDsBySlot           *shrinkingmap.ShrinkingMap[iotago.SlotIndex, map[iotago.TransactionID]struct{}]
```

Comment:
use `*memstorage.IndexedStorage[iotago.SlotIndex, iotago.TransactionID, struct{}]` instead

Reply:
Good idea, but I think it doesn't fit my use case exactly.